server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -118,29 +118,29 @@ public void addDynamic(String name, String prefix, Mapper mapper, DocumentParser
             // otherwise we strip off the first object path of the mapper name, load or create
             // the relevant object mapper, and then recurse down into it, passing the remainder
             // of the mapper name. So for a mapper 'foo.bar.baz', we locate 'foo' and then
-            // call addDynamic on it with the name 'bar.baz'.
+            // call addDynamic on it with the name 'bar.baz', and next call addDynamic on 'bar' with the name 'baz'.
             else {
                 int firstDotIndex = name.indexOf(".");
-                String childName = name.substring(0, firstDotIndex);
-                String fullChildName = prefix == null ? childName : prefix + "." + childName;
-                ObjectMapper.Builder childBuilder = findChild(fullChildName, context);
-                childBuilder.addDynamic(name.substring(firstDotIndex + 1), fullChildName, mapper, context);
-                add(childBuilder);
+                String parentName = name.substring(0, firstDotIndex);
+                String fullParentName = prefix == null ? parentName : prefix + "." + parentName;
+                ObjectMapper.Builder parentBuilder = findParentBuilder(fullParentName, context);
+                parentBuilder.addDynamic(name.substring(firstDotIndex + 1), fullParentName, mapper, context);
+                add(parentBuilder);
             }
         }
 
-        private static ObjectMapper.Builder findChild(String fullChildName, DocumentParserContext context) {
-            // does the child mapper already exist? if so, use that
-            ObjectMapper child = context.mappingLookup().objectMappers().get(fullChildName);
-            if (child != null) {
-                return child.newBuilder(context.indexSettings().getIndexVersionCreated());
+        private static ObjectMapper.Builder findParentBuilder(String fullName, DocumentParserContext context) {
+            // does the parent mapper already exist? if so, use that
+            ObjectMapper parent = context.mappingLookup().objectMappers().get(fullName);
+            if (parent != null) {
+                return parent.newBuilder(context.indexSettings().getIndexVersionCreated());
             }
-            // has the child mapper been added as a dynamic update already?
-            child = context.getDynamicObjectMapper(fullChildName);
-            if (child != null) {
-                return child.newBuilder(context.indexSettings().getIndexVersionCreated());
+            // has the parent mapper been added as a dynamic update already?
+            parent = context.getDynamicObjectMapper(fullName);
+            if (parent != null) {
+                return parent.newBuilder(context.indexSettings().getIndexVersionCreated());
             }
-            throw new IllegalArgumentException("Missing intermediate object " + fullChildName);
+            throw new IllegalStateException("Missing intermediate object " + fullName);
```

Comment:
this is the only relevant change in this file: IllegalArgumentException -> IllegalStateException . All the rest is renaming child to parent.